### PR TITLE
use batched gemm to speed up GPU and lower the JIT time

### PR DESCRIPTION
**Pull request by GiggleLiu**
Resolving #74 .

It can be merged after @Roger-luo tag the new release of BatchedRoutines.

Note: this is adapted from #2

### CODE REVIEW

**Comment by GiggleLiu** [comment]
# Benchmark
## Rule change
#### master branch
```
julia> OMEinsum.match_rule(ein"ijl,jkl->ikl")
OMEinsum.DefaultRule()

julia> OMEinsum.match_rule(ein"ijl,jkl->ik")
OMEinsum.PairWise()
```

#### new
```julia
julia> OMEinsum.match_rule(ein"ijl,jkl->ikl")
OMEinsum.BatchedContract()

julia> OMEinsum.match_rule(ein"ijl,jkl->ik")
OMEinsum.BatchedContract()
```

## benchmark - CPU
#### master
```julia
T = randn(100,100,1000);
julia> @benchmark ein"ijk,jlk->ilk"(T, T)
BenchmarkTools.Trial: 
  memory estimate:  76.31 MiB
  allocs estimate:  240
  --------------
  minimum time:     89.986 ms (9.70% GC)
  median time:      96.308 ms (8.90% GC)
  mean time:        109.259 ms (12.13% GC)
  maximum time:     207.974 ms (51.01% GC)
  --------------
  samples:          47
  evals/sample:     1

julia> @benchmark ein"ijk,jlk->il"(T, T)
BenchmarkTools.Trial: 
  memory estimate:  76.38 MiB
  allocs estimate:  238
  --------------
  minimum time:     133.128 ms (13.56% GC)
  median time:      152.739 ms (5.61% GC)
  mean time:        156.250 ms (7.65% GC)
  maximum time:     223.539 ms (45.41% GC)
  --------------
  samples:          32
  evals/sample:     1
```

#### new
```julia
julia> @benchmark ein"ijk,jlk->ilk"(T, T)
BenchmarkTools.Trial: 
  memory estimate:  76.30 MiB
  allocs estimate:  25
  --------------
  minimum time:     1.580 s (1.38% GC)
  median time:      1.611 s (2.59% GC)
  mean time:        1.606 s (2.59% GC)
  maximum time:     1.623 s (4.04% GC)
  --------------
  samples:          4
  evals/sample:     1

julia> @benchmark ein"ijk,jlk->il"(T, T)
BenchmarkTools.Trial: 
  memory estimate:  79.11 KiB
  allocs estimate:  22
  --------------
  minimum time:     81.214 ms (0.00% GC)
  median time:      85.310 ms (0.00% GC)
  mean time:        86.902 ms (0.00% GC)
  maximum time:     130.337 ms (0.00% GC)
  --------------
  samples:          58
  evals/sample:     1
```

## benchmark - GPU
#### master branch
```julia
julia> @benchmark (CuArrays.@sync ein"ijk,jlk->ilk"(T, T))
BenchmarkTools.Trial: 
  memory estimate:  10.50 KiB
  allocs estimate:  205
  --------------
  minimum time:     135.596 ms (0.00% GC)
  median time:      138.661 ms (0.00% GC)
  mean time:        139.749 ms (0.29% GC)
  maximum time:     153.708 ms (2.63% GC)
  --------------
  samples:          36
  evals/sample:     1

julia> @benchmark (CuArrays.@sync ein"ijk,jlk->il"(T, T))
BenchmarkTools.Trial: 
  memory estimate:  9.45 KiB
  allocs estimate:  185
  --------------
  minimum time:     1.873 s (0.00% GC)
  median time:      1.878 s (0.00% GC)
  mean time:        1.896 s (0.00% GC)
  maximum time:     1.938 s (0.00% GC)
  --------------
  samples:          3
  evals/sample:     1
```
#### new
```julia
julia> @benchmark (CuArrays.@sync ein"ijk,jlk->ilk"(T, T))
BenchmarkTools.Trial: 
  memory estimate:  15.48 KiB
  allocs estimate:  269
  --------------
  minimum time:     130.312 ms (0.00% GC)
  median time:      130.936 ms (0.00% GC)
  mean time:        132.623 ms (0.20% GC)
  maximum time:     143.398 ms (1.55% GC)
  --------------
  samples:          38
  evals/sample:     1

julia> @benchmark (CuArrays.@sync ein"ijk,jlk->il"(T, T))
BenchmarkTools.Trial: 
  memory estimate:  17.36 KiB
  allocs estimate:  320
  --------------
  minimum time:     131.481 ms (0.00% GC)
  median time:      134.130 ms (0.00% GC)
  mean time:        141.025 ms (0.19% GC)
  maximum time:     180.895 ms (1.67% GC)
  --------------
  samples:          36
  evals/sample:     1
```

## What can be improved
* the index information can be obtained statially by seperating function `analyse_gemm`, but I didn't do that to avoid potential compiling overhead.
* the performance of CPU version pairwise contraction falls back, but @Roger-luo said the performance is much better in the next version by approximately a factor of 2.

**Comment by Roger-luo** [comment]
I removed the conditional dependency hack in BatchedRoutines, use CuBatchedRoutines instead.

As rust-cargo says:

> the canonical way of implementing conditional dependencies is to have multiple packages

**Comment by GiggleLiu** [comment]
> I removed the conditional dependency hack in BatchedRoutines, use CuBatchedRoutines instead.
> 
> As rust-cargo says:
> 
> > the canonical way of implementing conditional dependencies is to have multiple packages

This pr uses the beached routines in CuArrays. It that correct?

**Comment by Roger-luo** [comment]
yes, you can wrap up that routine in cublas yourself, but the problem is CuBatchedRoutines provides a unified API simply load `CuBatchedRoutines` with `Require` will do the job (in principal you don't need to write anything else). I can register `CuBatchedRoutines` a bit later.

And there're a few other routines which is not in cublas that I think you will also want to dispatch, e.g `batched_tr` etc.

File: src/EinRule.jl [hunk 1]
```diff
@@ -83,6 +84,17 @@ function match_rule(::Type{Identity}, ixs, iy)
     ixs === (iy,) && allunique(iy)
 end
 
+function match_rule(::Type{BatchedContract}, ixs::NTuple{X, NTuple{N,T} where N}, iy::NTuple{M,T}) where {T,X,M}
```
Comment (under-Peter) [inline src/EinRule.jl]: I'd suggest replacing the function body with 
```julia
X == 2 || return false
iA, iB = ixs
all(allunique, (iA, iB, iy)) || return false
allinds = (iA..., iB..., iy...)
all(x -> count(==(x), allinds) > 1, allinds)
```
which is about a ~10 times faster for small tuples than the dictionary version, probably due to both the dict and `tunique` allocating.

File: src/utils.jl [hunk 2]
```diff
@@ -59,4 +59,20 @@ julia> allunique((1,2,3,1))
 false
 ```
 """
```
Comment (under-Peter) [inline src/utils.jl]: Not sure this is necessary, see
```
julia> @btime allunique_old((1,2,3,4,5,6,7,8,9,10));
       @btime allunique_old((1,2,2,3,4,5));
       @btime allunique_old((1,2,3));
       @btime allunique_new((1,2,3,4,5,6,7,8,9,10));
       @btime allunique_new((1,2,2,3,4,5));
       @btime allunique_new((1,2,3));
  20.945 ns (0 allocations: 0 bytes) #old
  1.811 ns (0 allocations: 0 bytes) #old
  5.032 ns (0 allocations: 0 bytes) #old
  26.312 ns (0 allocations: 0 bytes) #new
  9.006 ns (0 allocations: 0 bytes) #new
  1.811 ns (0 allocations: 0 bytes) #new
```
So sometimes it's better, sometimes it's worse - negligible times anyway but imho the one-liner reads cleaner.

File: src/einsequence.jl [hunk 3]
```diff
@@ -202,3 +202,10 @@ end
 
 extractxs(xs, x::NestedEinsumStable) = x(xs...)
 extractxs(xs, i::Int) = xs[i]
+
+
+function match_rule(code::NestedEinsumStable)
```
Comment (under-Peter) [inline src/einsequence.jl]: what's happening here? Does that change anything - is it ever used?

File: src/batched_contract.jl [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+# batched, dynamic version
+function batched_contract(iAs, A::AbstractArray, iBs, B::AbstractArray, iOuts::NTuple{NO,T}) where {NO,T}
+    pA, sAb, sAs, sAp, pB, sBs, sBb, sBp, sAB, pOut = analyse_batched(iAs, size(A), iBs, size(B), iOuts)
+
+    A, B = align_eltypes(A, B)
+    Apr = reshape(conditioned_permutedims(A, pA), sAb, sAs, sAp)
+    Bpr = reshape(conditioned_permutedims(B, pB), sBs, sBb, sBp)
+    AB = _batched_gemm('N','N', Apr, Bpr)
+    AB = conditioned_permutedims(reshape(AB, sAB...), [pOut...])
```
Comment (under-Peter) [inline src/batched_contract.jl]: Use `collect(pOut)` instead of `[pOut...]`

File: src/batched_contract.jl [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# batched routines
+@inline indexpos(iAs, i)::Int = findfirst(==(i), iAs)
+
+# can be used in either static or dynamic invoke
+function analyse_batched(iAs, sA, iBs, sB, iOuts)
```
Comment (under-Peter) [inline src/batched_contract.jl]: I'll have to go over this when I got more time on my hand.

File: src/einsequence.jl (hunk 3 above)
Comment (GiggleLiu) [reply src/einsequence.jl]: No, it is just a debugging utility. You can use it to see the rules used in a nested einsum code.
Before this, only `match_rule(ein"ij,jk->ik")` is defined, but  `match_rule(ein"(ij,jk),kl->il")` errors. I should have writen a test for it.

File: src/batched_contract.jl (hunk 4 above)
Comment (GiggleLiu) [reply src/batched_contract.jl]: They are equivalent?

```julia
julia> @benchmark collect($tp)
BenchmarkTools.Trial: 
  memory estimate:  112 bytes
  allocs estimate:  1
  --------------
  minimum time:     35.288 ns (0.00% GC)
  median time:      36.577 ns (0.00% GC)
  mean time:        49.557 ns (18.40% GC)
  maximum time:     45.822 μs (99.83% GC)
  --------------
  samples:          10000
  evals/sample:     992

julia> @benchmark [$tp...]
BenchmarkTools.Trial: 
  memory estimate:  112 bytes
  allocs estimate:  1
  --------------
  minimum time:     30.851 ns (0.00% GC)
  median time:      32.208 ns (0.00% GC)
  mean time:        45.245 ns (20.71% GC)
  maximum time:     44.979 μs (99.88% GC)
  --------------
  samples:          10000
  evals/sample:     993
```

File: src/batched_contract.jl (hunk 5 above)
Comment (GiggleLiu) [reply src/batched_contract.jl]: I didn't change much from your original code in PR #2 .

File: src/utils.jl (hunk 2 above)
Comment (GiggleLiu) [reply src/utils.jl]: Ok, i see. will change it back. :+1:

File: src/EinRule.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/EinRule.jl]: Good, looks correct and shorter!

**Comment by Roger-luo** [comment]
use the lastest release, I removed CUDA related hack in the package. It had some issue on CI before.

**Comment by GiggleLiu** [comment]
> use the lastest release, I removed CUDA related hack in the package. It had some issue on CI before.

:+1: Thanks, having changed the compat, now the tests pass!

File: src/batched_contract.jl (hunk 4 above)
Comment (under-Peter) [reply src/batched_contract.jl]: IMO `collect` is the canonical way, although splatting works as well. Not too important and certainly no bottleneck.

**Comment by under-Peter** [comment]
@GiggleLiu , ok with me to merge this - ok with you?

**Comment by GiggleLiu** [comment]
good, please. :+1:

**Pull request merged**